crates/hir_ty/src/tests/method_resolution.rs

```diff
@@ -1051,3 +1051,41 @@ fn dyn_trait_super_trait_not_in_scope() {
         "#]],
     );
 }
+
+#[test]
+fn method_resolution_foreign_opaque_type() {
+    check_infer(
+        r#"
+        extern "C" {
+            type S;
+            fn f() -> &'static S;
+        }
+
+        impl S {
+            fn foo(&self) -> bool {
+                true
+            }
+        }
+
+        fn test() {
+            let s = unsafe { f() };
+            s.foo();
+        }
+        "#,
+        // FIXME: 's.foo()' should be `bool`.
```

Comment:
hmm why isn't it? inherent method resolution doesn't even go through Chalk.

Reply:
Yes, i found the problem! See the newer commit.

Reply:
Aah saw your fix, that makes sense.